src/storage/mod.rs

```diff
@@ -591,11 +847,52 @@ pub trait UnprotectedStorage<T>: TryDefault {
     ///
     /// May only be called if an element with `id` was `insert`ed and not yet
     /// removed / dropped.
+    ///
+    /// Caller must ensure this is cleared from the mask even if the drop impl
+    /// of the component panics and this unwinds. Usually, this can be
+    /// accomplished by removing the id from the mask just before calling this.
     unsafe fn drop(&mut self, id: Index) {
-        self.remove(id);
+        // SAFETY: Requirements passed to the caller.
+        unsafe { self.remove(id) };
     }
 }
 
+/// Used by the framework to mutably access components in contexts where
+/// exclusive access to the storage is not possible.
+pub trait SharedGetMutStorage<T>: UnprotectedStorage<T> {
```

Comment:
The base `UnprotectedStorage` trait means that we can implement `LendJoin` for it. `SharedGetMutStorage` enables implementing `Join`. And finally `DistinctStorage` enables `ParJoin`.